### PR TITLE
KVStore: Reduce lock contention in `RegionPersister::doPersist`

### PR DESCRIPTION
**Pull request by JaySon-Huang**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tiflash/issues/8583

Problem Summary:

https://github.com/pingcap/tiflash/blob/5980908e68629f40c2191ae2d100f41d554dce44/dbms/src/Storages/KVStore/MultiRaft/RegionPersister.cpp#L99-L125

`RegionPersister::doPersist` acquires a lock and calls `PageStorage::write` with the lock. This means when we're persisting Region meta, all other threads can not do the persistence for other Region meta. Which is not reasonable.

### What is changed and how it works?

There is a `region_task_lock` to ensure that only one thread could call `RegionPersister::doPersist` at the same time. And `PageStorage` support multi-thread reading and writing for different page_id. There is no need to acquire another lock in `RegionPersister::doPersist` that could slowdown the raft apply process

And also remove calling `KVStore::persistRegion` with a null `region_task_lock`. So we can simplify codes and remove the member `RegionPersister:: RegionManager`

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Reduce the impact of disk performance jitter on write throughput
```

### CODE REVIEW

File: dbms/src/Storages/KVStore/KVStore.cpp [hunk 1]
```diff
@@ -375,33 +374,25 @@ void KVStore::setRegionCompactLogConfig(UInt64 rows, UInt64 bytes, UInt64 gap, U
...
+    auto reason_id = magic_enum::enum_underlying(reason);
+    std::string caller = fmt::format("{} {}", PersistRegionReasonMap[reason_id], extra_msg);
+    LOG_INFO(
+        log,
+        "Start to persist {}, cache size: {} bytes for `{}`",
+        region.getDebugString(),
+        region.dataSize(),
+        caller);
+    region_persister->persist(region, region_task_lock);
+    LOG_DEBUG(log, "Persist {} done, cache size: {} bytes", region.toString(false), region.dataSize());
```
Comment (CalvinNeo) [inline dbms/src/Storages/KVStore/KVStore.cpp:395]: Use `region size` for `region.dataSize()`. Though it's quite misleading that we have both `approx_mem_cache_bytes` and `region.dataSize()`. I think maybe we optimize it later.

**Review by CalvinNeo — APPROVED** [review]
lgtm

**Review by Lloyd-Pottiger — APPROVED** [review]

**Comment by JaySon-Huang** [comment]
Change the release note from "write throughput" to "read latency"

Tests with some failpoint injects (99%-10ms and %0.7-100ms and %0.2-1000ms) latency on `RegionPersister::doPersist` show that the changes mainly affect the read latency caused by `wait index` but not the write throughput

image

**Pull request merged**